File: lib/data-protection.ts
import fs from "node:fs"
import path from "node:path"
import crypto from "node:crypto"

const DATA_DIR = process.env.DATA_DIR?.trim() || path.join(process.cwd(), "data")
const DB_PATH = path.join(DATA_DIR, "santa-luzia.json")
const BACKUP_DIR = path.join(DATA_DIR, "backups-santa-luzia")
const HEALTH_PATH = path.join(DATA_DIR, "database-health.json")
const MAX_BACKUPS = 8
const WATCH_INTERVAL_MS = 2_000
const SNAPSHOT_DEBOUNCE_MS = 900

type DatabaseHealth = {
  checkedAt: string
  status: "ok" | "recovered" | "missing" | "error"
  databasePath: string
  size: number
  sha256: string | null
  backupCount: number
  lastBackup: string | null
  recoveredFrom?: string | null
  error?: string | null
}

type GlobalProtection = typeof globalThis & {
  __santaLuziaDataProtection?: {
    started: boolean
    timer?: NodeJS.Timeout
    lastSha?: string
  }
}

function ensureDirectories() {
  fs.mkdirSync(DATA_DIR, { recursive: true })
  fs.mkdirSync(BACKUP_DIR, { recursive: true })
}

function sha256(buffer: Buffer) {
  return crypto.createHash("sha256").update(buffer).digest("hex")
}

function readValidDatabase(file: string) {
  if (!fs.existsSync(file)) return null
  const buffer = fs.readFileSync(file)
  if (!buffer.length) throw new Error(`Banco vazio: ${file}`)
  const parsed = JSON.parse(buffer.toString("utf8")) as Record<string, unknown>
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error(`Estrutura inválida: ${file}`)
  const expectedArrays = [
    "usuarios",
    "registros",
    "escalas",
    "formacoes",
    "formacao_presencas",
    "ranking_ajustes",
  ]
  for (const key of expectedArrays) {
    if (key in parsed && !Array.isArray(parsed[key])) throw new Error(`Campo ${key} não é uma lista válida.`)
  }
  return { buffer, parsed, sha: sha256(buffer), size: buffer.length }
}

function backupFiles() {
  ensureDirectories()
  return fs.readdirSync(BACKUP_DIR)
    .filter((name) => /^santa-luzia-\d{8}-\d{6}-[a-f0-9]{12}\.json$/.test(name))
    .map((name) => ({ name, file: path.join(BACKUP_DIR, name), mtime: fs.statSync(path.join(BACKUP_DIR, name)).mtimeMs }))
    .sort((a, b) => b.mtime - a.mtime)
}

function trimBackups() {
  const files = backupFiles()
  for (const item of files.slice(MAX_BACKUPS)) {
    try { fs.rmSync(item.file, { force: true }) } catch {}
  }
}

function timestamp(date = new Date()) {
  const p = (value: number) => String(value).padStart(2, "0")
  return `${date.getFullYear()}${p(date.getMonth() + 1)}${p(date.getDate())}-${p(date.getHours())}${p(date.getMinutes())}${p(date.getSeconds())}`
}

function atomicWrite(file: string, buffer: Buffer) {
  const temp = `${file}.tmp-${process.pid}-${Date.now()}`
  const fd = fs.openSync(temp, "w", 0o600)
  try {
    fs.writeFileSync(fd, buffer)
    fs.fsyncSync(fd)
  } finally {
    fs.closeSync(fd)
  }
  fs.renameSync(temp, file)
  try {
    const dirFd = fs.openSync(path.dirname(file), "r")
    try { fs.fsyncSync(dirFd) } finally { fs.closeSync(dirFd) }
  } catch {}
}

function writeHealth(health: DatabaseHealth) {
  try { atomicWrite(HEALTH_PATH, Buffer.from(`${JSON.stringify(health, null, 2)}\n`, "utf8")) } catch {}
}

function latestValidBackup() {
  for (const item of backupFiles()) {
    try {
      const valid = readValidDatabase(item.file)
      if (valid) return { ...item, ...valid }
    } catch {}
  }
  return null
}

function createBackup(force = false) {
  if (!fs.existsSync(DB_PATH)) return null
  const current = readValidDatabase(DB_PATH)
  if (!current) return null
  const backups = backupFiles()
  if (!force && backups.length) {
    try {
      const last = readValidDatabase(backups[0].file)
      if (last?.sha === current.sha) return { file: backups[0].file, sha: current.sha, size: current.size, reused: true }
    } catch {}
  }
  const name = `santa-luzia-${timestamp()}-${current.sha.slice(0, 12)}.json`
  const target = path.join(BACKUP_DIR, name)
  atomicWrite(target, current.buffer)
  trimBackups()
  return { file: target, sha: current.sha, size: current.size, reused: false }
}

function recoverIfNeeded() {
  ensureDirectories()
  if (!fs.existsSync(DB_PATH)) {
    const backup = latestValidBackup()
    if (!backup) {
      writeHealth({ checkedAt: new Date().toISOString(), status: "missing", databasePath: DB_PATH, size: 0, sha256: null, backupCount: backupFiles().length, lastBackup: null })
      return { status: "missing" as const, sha: "" }
    }
    atomicWrite(DB_PATH, backup.buffer)
    writeHealth({ checkedAt: new Date().toISOString(), status: "recovered", databasePath: DB_PATH, size: backup.size, sha256: backup.sha, backupCount: backupFiles().length, lastBackup: path.basename(backup.file), recoveredFrom: path.basename(backup.file) })
    return { status: "recovered" as const, sha: backup.sha }
  }

  try {
    const current = readValidDatabase(DB_PATH)!
    const backup = createBackup(false)
    writeHealth({ checkedAt: new Date().toISOString(), status: "ok", databasePath: DB_PATH, size: current.size, sha256: current.sha, backupCount: backupFiles().length, lastBackup: backup ? path.basename(backup.file) : null })
    return { status: "ok" as const, sha: current.sha }
  } catch (error) {
    const corrupt = `${DB_PATH}.corrompido-${timestamp()}.json`
    try { fs.renameSync(DB_PATH, corrupt) } catch {}
    const backup = latestValidBackup()
    if (!backup) {
      writeHealth({ checkedAt: new Date().toISOString(), status: "error", databasePath: DB_PATH, size: 0, sha256: null, backupCount: backupFiles().length, lastBackup: null, error: error instanceof Error ? error.message : String(error) })
      return { status: "error" as const, sha: "" }
    }
    atomicWrite(DB_PATH, backup.buffer)
    writeHealth({ checkedAt: new Date().toISOString(), status: "recovered", databasePath: DB_PATH, size: backup.size, sha256: backup.sha, backupCount: backupFiles().length, lastBackup: path.basename(backup.file), recoveredFrom: path.basename(backup.file), error: error instanceof Error ? error.message : String(error) })
    console.warn(`[Proteção do banco] Banco principal inválido. Recuperado de ${path.basename(backup.file)}.`)
    return { status: "recovered" as const, sha: backup.sha }
  }
}

function snapshotAfterWrite(state: NonNullable<GlobalProtection["__santaLuziaDataProtection"]>) {
  if (state.timer) clearTimeout(state.timer)
  state.timer = setTimeout(() => {
    try {
      const current = readValidDatabase(DB_PATH)
      if (!current || current.sha === state.lastSha) return
      const backup = createBackup(false)
      state.lastSha = current.sha
      writeHealth({ checkedAt: new Date().toISOString(), status: "ok", databasePath: DB_PATH, size: current.size, sha256: current.sha, backupCount: backupFiles().length, lastBackup: backup ? path.basename(backup.file) : null })
    } catch (error) {
      writeHealth({ checkedAt: new Date().toISOString(), status: "error", databasePath: DB_PATH, size: 0, sha256: null, backupCount: backupFiles().length, lastBackup: backupFiles()[0]?.name ?? null, error: error instanceof Error ? error.message : String(error) })
    }
  }, SNAPSHOT_DEBOUNCE_MS)
  state.timer.unref?.()
}

export function iniciarProtecaoDadosSantaLuzia() {
  const globals = globalThis as GlobalProtection
  if (globals.__santaLuziaDataProtection?.started) return
  const state: NonNullable<GlobalProtection["__santaLuziaDataProtection"]> = { started: true }
  globals.__santaLuziaDataProtection = state

  const recovery = recoverIfNeeded()
  state.lastSha = recovery.sha

  fs.watchFile(DB_PATH, { interval: WATCH_INTERVAL_MS, persistent: false }, (current, previous) => {
    if (current.mtimeMs === previous.mtimeMs && current.size === previous.size) return
    snapshotAfterWrite(state)
  })

  process.once("beforeExit", () => {
    try { createBackup(false) } catch {}
  })
}

export function lerSaudeBancoSantaLuzia(): DatabaseHealth | null {
  try { return JSON.parse(fs.readFileSync(HEALTH_PATH, "utf8")) as DatabaseHealth } catch { return null }
}

File: python-backend/santa_luzia_backend/data_protection.py
from __future__ import annotations

import hashlib
import json
import os
import shutil
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .store import DATA_DIR, DB_PATH

BACKUP_DIR = DATA_DIR / "backups-santa-luzia"
HEALTH_PATH = DATA_DIR / "database-health.json"
MAX_BACKUPS = 8
_lock = threading.RLock()
_last_sha: str | None = None


def _sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _valid_database(path: Path):
    if not path.exists():
        return None
    data = path.read_bytes()
    if not data:
        raise ValueError(f"Banco vazio: {path}")
    parsed = json.loads(data.decode("utf-8"))
    if not isinstance(parsed, dict):
        raise ValueError(f"Estrutura inválida: {path}")
    for key in ("usuarios", "registros", "escalas", "formacoes", "formacao_presencas", "ranking_ajustes"):
        if key in parsed and not isinstance(parsed[key], list):
            raise ValueError(f"Campo {key} não é uma lista válida.")
    return {"data": data, "sha": _sha256(data), "size": len(data)}


def _backup_files():
    BACKUP_DIR.mkdir(parents=True, exist_ok=True)
    items = [item for item in BACKUP_DIR.iterdir() if item.is_file() and item.name.startswith("santa-luzia-") and item.suffix == ".json"]
    return sorted(items, key=lambda item: item.stat().st_mtime, reverse=True)


def _write_health(value: dict[str, Any]):
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    tmp = HEALTH_PATH.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    os.replace(tmp, HEALTH_PATH)


def _health(status: str, current: dict[str, Any] | None, *, recovered_from: str | None = None, error: str | None = None):
    backups = _backup_files()
    value = {
        "checkedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "status": status,
        "databasePath": str(DB_PATH),
        "size": int(current["size"]) if current else 0,
        "sha256": current["sha"] if current else None,
        "backupCount": len(backups),
        "lastBackup": backups[0].name if backups else None,
        "recoveredFrom": recovered_from,
        "error": error,
    }
    _write_health(value)
    return value


def protect_database() -> dict[str, Any]:
    global _last_sha
    with _lock:
        DATA_DIR.mkdir(parents=True, exist_ok=True)
        BACKUP_DIR.mkdir(parents=True, exist_ok=True)
        if not DB_PATH.exists():
            for backup in _backup_files():
                try:
                    valid = _valid_database(backup)
                    if valid:
                        shutil.copyfile(backup, DB_PATH)
                        _last_sha = valid["sha"]
                        return _health("recovered", valid, recovered_from=backup.name)
                except Exception:
                    continue
            return _health("missing", None)
        try:
            current = _valid_database(DB_PATH)
            if not current:
                return _health("missing", None)
            if current["sha"] != _last_sha:
                backups = _backup_files()
                reuse = False
                if backups:
                    try:
                        latest = _valid_database(backups[0])
                        reuse = bool(latest and latest["sha"] == current["sha"])
                    except Exception:
                        reuse = False
                if not reuse:
                    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
                    target = BACKUP_DIR / f"santa-luzia-{stamp}-{current['sha'][:12]}.json"
                    target.write_bytes(current["data"])
                    for old in _backup_files()[MAX_BACKUPS:]:
                        old.unlink(missing_ok=True)
                _last_sha = current["sha"]
            return _health("ok", current)
        except Exception as error:
            corrupt = DB_PATH.with_name(f"santa-luzia.json.corrompido-{datetime.now().strftime('%Y%m%d-%H%M%S')}")
            try:
                os.replace(DB_PATH, corrupt)
            except Exception:
                pass
            for backup in _backup_files():
                try:
                    valid = _valid_database(backup)
                    if valid:
                        shutil.copyfile(backup, DB_PATH)
                        _last_sha = valid["sha"]
                        return _health("recovered", valid, recovered_from=backup.name, error=str(error))
                except Exception:
                    continue
            return _health("error", None, error=str(error))


def read_health():
    try:
        value = json.loads(HEALTH_PATH.read_text(encoding="utf-8"))
        return value if isinstance(value, dict) else None
    except Exception:
        return None
